### PR TITLE
Feature/add default credentials provider

### LINKED ISSUE #290: [FEATURE] Default getCredentials for AWS Sigv4 Signer
### Is your feature request related to a problem?

In https://github.com/opensearch-project/opensearch-js/pull/279 we introduced Sigv4 signing. To use this one must decide which version of AWS SDK to import, then implement different ways to get credentials depending on the version.

### What solution would you like?

A version-agonostic default implementation. I want to be able to write the following regardless of the version of AWS SDK I am importing and without having to include any boilerplate for getting/refreshing credentials. And I want it to "just work" (TM) when I set, for example, credentials via ENV AWS_ variables.

```js
const client = new Client({
  ...AwsSigv4Signer({
      region: 'us-east-1',
      node: "https://search-xxx.region.es.amazonaws.com",
  })
});
```

(this is similar to how one can use a client in the [Java SDK](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/request-signing.html#request-signing-java)).

### PR DESCRIPTION
**Pull request by rawpixel-vincent**
### Description
fix #290 but I don't like it unless someone comes with a better way to resolve the SDK

(includes #294 cause I needed it to run the tests)

Tested the dynamic import with both version of aws sdk and tested the integration with aws access key and aws role with the sdk v2

### CODE REVIEW

File: lib/aws/AwsSigv4Signer.js [hunk 1]
```diff
@@ -15,6 +15,50 @@ const Transport = require('../Transport');
...
+  } catch (err) {
+    // Ignore
+  }
+
+  throw new AwsSigv4SignerError(
+    'Unable to find a valid AWS SDK, please provide a valid getCredentials function to AwsSigv4Signer options.'
+  );
+};
+
+const defaultCredentialsAcquirer = () =>
```
Comment (dblock) [inline lib/aws/AwsSigv4Signer.js:51]: Should this be an `awsDefaultCredentialsProvider()`?

File: test/unit/lib/aws/awssigv4signer.test.js [hunk 2]
```diff
@@ -77,20 +77,44 @@ test('Sign with SigV4 failure (with empty region)', (t) => {
   }
 });
 
-test('Sign with SigV4 failure (without getCredentials function)', (t) => {
```
Comment (dblock) [inline test/unit/lib/aws/awssigv4signer.test.js:80]: I think you want both tests, without getCredentials and with default.

File: lib/aws/AwsSigv4Signer.js (hunk 1 above)
Comment (rawpixel-vincent) [reply lib/aws/AwsSigv4Signer.js:51]: roger that

File: test/unit/lib/aws/awssigv4signer.test.js (hunk 2 above)
Comment (rawpixel-vincent) [reply test/unit/lib/aws/awssigv4signer.test.js:80]: the default is without getCredentials, the others test use custom provided getCredentials function

File: lib/aws/AwsSigv4Signer.js (hunk 1 above)
Comment (rawpixel-vincent) [reply lib/aws/AwsSigv4Signer.js]: done

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
I've restarted some flakey tests here and they passed. If someone wants to contribute to less flakeness that would be appreciated.

**Comment by harshavamsi** [comment]
> I've restarted some flakey tests here and they passed. If someone wants to contribute to less flakeness that would be appreciated.

@nhtruong is working on #268. Looks like a race condition.

**Comment by dblock** [comment]
@VachaShah you're good with this?

File: package.json [hunk 3]
```diff
@@ -34,16 +34,16 @@
     "index"
   ],
   "scripts": {
-    "test": "npm run lint && tap test/{unit,acceptance}/{*,**/*}.test.js && npm run test:types",
-    "test:unit": "tap test/unit/{*,**/*}.test.js",
+    "test": "npm run lint && tap test/{unit,acceptance}/{*,**/*,**/**/*}.test.js && npm run test:types",
```
Comment (VachaShah) [inline package.json:37]: Looks like this already got merged in PR #294?

**Review by VachaShah — APPROVED** [review]
LGTM! Just the comment about `package.json` changes.

File: package.json (hunk 3 above)
Comment (rawpixel-vincent) [reply package.json:37]: yes, I included that commit cause I needed this command to run the tests in this branch.

see in the PR description:

> (includes https://github.com/opensearch-project/opensearch-js/pull/294 cause I needed it to run the tests)

Comment (rawpixel-vincent) [reply package.json]: rebased with main 👌🏻

**Comment by rawpixel-vincent** [comment]
rebased with main

**Comment by rawpixel-vincent** [comment]
oups

**Pull request merged**